Assets/Labrys/FeatureEditor/FeatureAsset.cs

```diff
@@ -0,0 +1,372 @@
+		private List<Link> linkData;
+
+		[SerializeField, HideInInspector]
+		private List<Vector2Int> selectionPositions;
+
+		public void OnAfterDeserialize()
+		{
+			if (sectionPositions.Count != sectionData.Count
+				|| linkPositions.Count != linkData.Count)
+				throw new ArgumentException("Malformed feature data");
+
+			for(int i = 0; i < sectionPositions.Count; i++)
+			{
+				sections.Add(sectionPositions[i], sectionData[i]);
+			}
+
+			for (int i = 0; i < linkPositions.Count; i++)
+			{
+				links.Add(linkPositions[i], linkData[i]);
+			}
+
+			for(int i = 0; i < selectionPositions.Count; i++)
+			{
+				selected.Add(selectionPositions[i]);
+			}
+		}
+
+		public void OnBeforeSerialize()
+		{
+			sectionPositions.Clear();
+			sectionData.Clear();
+			foreach(KeyValuePair<Vector2Int, Section> kvp in sections)
+			{
+				sectionPositions.Add(kvp.Key);
+				sectionData.Add(kvp.Value);
+			}
+
+			linkPositions.Clear();
+			linkData.Clear();
+			foreach (KeyValuePair<Vector2Int, Link> kvp in links)
+			{
+				linkPositions.Add(kvp.Key);
+				linkData.Add(kvp.Value);
+			}
+
+			selectionPositions.Clear();
+			foreach(Vector2Int position in selected)
+			{
+				selectionPositions.Add(position);
+			}
+		}
+		#endregion
+
+		public FeatureAsset()
+		{
+			sections = new Dictionary<Vector2Int, Section>();
+			links = new Dictionary<Vector2Int, Link>();
+			selected = new HashSet<Vector2Int>();
+
+			sectionPositions = new List<Vector2Int>();
+			sectionData = new List<Section>();
+			linkPositions = new List<Vector2Int>();
+			linkData = new List<Link>();
+			selectionPositions = new List<Vector2Int>();
+		}
+
+		public void AddSection(Vector2Int gridPosition)
+		{
+			if(Section.IsValidPosition(gridPosition)
+				&& !sections.ContainsKey(gridPosition))
+			{
+				sections.Add(gridPosition, new Section() { Variant = "default" });
+				UpdateLinks(gridPosition);
+			}
+		}
+
+		public bool RemoveSection(Vector2Int gridPosition)
+		{
+			if(sections.Remove(gridPosition))
+			{
+				UpdateLinks(gridPosition);
+				DeselectSection(gridPosition);
+				return true;
+			}
+			return false;
+		}
+
+		public bool SelectSection(Vector2Int gridPosition)
+		{
+			if(Section.IsValidPosition(gridPosition))
+			{
+				return selected.Add(gridPosition);
+			}
+			return false;
+		}
+
+		public bool IsSelected(Vector2Int gridPosition)
+		{
+			if (Section.IsValidPosition(gridPosition))
+			{
+				return selected.Contains(gridPosition);
+			}
+			return false;
+		}
+
+		public bool DeselectSection(Vector2Int gridPosition)
+		{
+			return selected.Remove(gridPosition);
+		}
+
+		public bool HasSectionAt(Vector2Int gridPosition)
+		{
+			return Section.IsValidPosition(gridPosition) 
+				&& sections.ContainsKey(gridPosition);
+		}
+
+		public bool HasLinkAt(Vector2Int gridPosition)
+		{
+			return !Section.IsValidPosition(gridPosition) 
+				&& links.ContainsKey(gridPosition);
+		}
+
+		public bool TryGetSection(Vector2Int gridPosition, out Section section)
```

Comment:
petition to replace this with regular `GetSection` method. The `bool` return type is never used, only the value of `section`. Additionally, because of the restrictions on `AddSection` and `GetSections`, it's impossible to add a null Section to the internal lists, so there won't be a case where we'd return `true` and `section` == null.

Reply:
Return value is used on line 109 in `FromFeature()`.  Also, you are correct that the `out` parameter `section` will never be null, the name and syntax of `TryGetSection()` flows better in cases where a valid section cannot be returned.

Currently, a retrieval looks like this:
```cs
if (TryGetSection(position, out Section section)) {
    // do something with section
} else {
    // no section at position, do something else
}
```

Your proposed change (if I understand correctly) would result in something like this:
```cs
Section section = GetSection(position);
if (section != null) {
    // do stuff with section
} else {
    // there ain't no section; do other stuff
}
```
Which isn't horrible, but not as clear, imo.